### PR TITLE
metadce: JS function support

### PR DESCRIPTION
**Pull request by kripken**
This adds JS function scanning support to metadce. That means that we can DCE the combined JS+wasm in a way that doesn't treat the entire JS world as a single DCE entity (which we did before), instead we see individual functions, and can collect subgraphs between them and wasm.

An example of something this can clean up are the dynCall/invoke functions. wasm calls invoke in JS, which does a try-catch, which calls back into wasm using dynCall. With this PR, if an invoke is not used, we can see that the dynCall is not used either, and both are removable.

Even better, this can remove runtime methods like `abort()` that are imported into wasm, and wasm runtime exports like `stackAlloc()` that are exported from wasm. With this PR, if you compile something minimal like
```
#include <emscripten.h>

EMSCRIPTEN_KEEPALIVE
int add(int x, int y) {
  return x;
}
```
then you get something pretty minimal - a 132 byte wasm file! (which still contains a little unnecessary stuff, see #5966)

On more C/C++ hello-world type examples, where the runtime is less noticeable anyhow, this typically shrinks by a few %, e.g. 4% on libc++ hello world's JS.

This PR is mostly straightforward, the one tricky area are dynCalls, since we want to support `dynCall('v', ptr)`, i.e., giving the signature at runtime. This PR will analyze those, if it sees a static string being passed to `dynCall`, it will consider it a static dynCall, but if it sees something it cannot prove is static, it will add links to all the `dynCall_*` methods. This happens in practice if you use embind or addFunction, for example.

This PR also adds a test for the wasm backend, if the wasm backend is not the default anyhow, and if fastcomp was built with it. That's the case for me locally, but not the case on the emsdk, so this is skipped on the bots. But it helps me locally, and perhaps we should start to build the wasm backend with fastcomp (maybe after the next llvm merge, though).

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -921,6 +916,12 @@ def check(input_file):
       # Note the exports the user requested
       shared.Building.user_requested_exports = shared.Settings.EXPORTED_FUNCTIONS[:]
```
Comment (kripken) [inline emcc.py]: this change just moves the code to the right place - it confused me when debugging, before

File: src/embind/embind.js [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 // -- jshint doesn't understand library syntax, so we need to specifically tell it about the symbols we define
 /*global typeDependencies, flushPendingDeletes, getTypeName, getBasestPointer, throwBindingError, UnboundTypeError, _embind_repr, registeredInstances, registeredTypes, getShiftFromSize*/
-/*global ensureOverloadTable, requireFunction, awaitingDependencies, makeLegalFunctionName, embind_charCodes:true, registerType, createNamedFunction, RegisteredPointer, throwInternalError*/
+/*global ensureOverloadTable, embind__requireFunction, awaitingDependencies, makeLegalFunctionName, embind_charCodes:true, registerType, createNamedFunction, RegisteredPointer, throwInternalError*/
```
Comment (kripken) [inline src/embind/embind.js]: the embind changes are just renamings - requireFunction is special in that it creates dynamic dynCalls, and the name confused me when debugging, so I added embind in the name to clarify what it's doing.

File: tests/test_other.py [hunk 3]
```diff
@@ -7730,38 +7738,69 @@ def break_cashew():
...
-        assert not_exists not in sent, [not_exists, sent]
-      wasm_size = os.stat('a.out.wasm').st_size
-      ratio = (wasm_size - expected_wasm_size) / float(expected_wasm_size)
-      assert ratio < 0.05, [expected_wasm_size, wasm_size, ratio]
+    def test(filename, expectations):
+      sizes = {}
+      # in -Os, -Oz, we remove imports wasm doesn't need
+      for args, expected_len, expected_exists, expected_not_exists, expected_wasm_size, expected_wasm_imports, expected_wasm_exports in expectations:
+        print(args, expected_len, expected_exists, expected_not_exists, expected_wasm_size, expected_wasm_imports, expected_wasm_exports)
+        subprocess.check_call([PYTHON, EMCC, filename] + args + ['-s', 'WASM=1', '-g2'])
```
Comment (dschuff) [inline tests/test_other.py]: convert to run_process?

Comment (kripken) [reply tests/test_other.py]: Fixed, thanks.

**Pull request merged**